### ISSUE TITLE
Current line is not highlighed

### ISSUE THREAD

**Issue by habamax** [liuchengxu/vim-clap#130]
**Environment (please complete the following information):**
 - OS: windows 10
 - (Neo)Vim version: 8.1.2318
 - vim-clap version: latest master
 - Have you reproduced with a minimal vimrc: yes

**Describe the bug**
With `C-k` and `C-j` current list item is not visually selected

**To Reproduce**
- Open vim
- Clap files
- `C-j`

1. Create the minimal vimrc `min.vim`:

```vim
set nocompatible
set runtimepath^=/path/to/vim-clap
syntax on
filetype plugin indent on
```

**Expected behavior**
current line should be highlighted

**Screenshots**
![image](https://user-images.githubusercontent.com/234774/69152568-bea17280-0aed-11ea-8169-af43d8e67de2.png)

**Comment by habamax** [comment]
![vim-clap-line](https://user-images.githubusercontent.com/234774/69152809-2ce63500-0aee-11ea-8f4b-8d4738b9b6c8.gif)

**Comment by liuchengxu** [comment]
I did find a related problem when checking this issue. But the `<C-K>` and `<C-J>` works well on my side.

```diff
diff --git a/autoload/clap/impl.vim b/autoload/clap/impl.vim
index 7da1115..1a42111 100644
--- a/autoload/clap/impl.vim
+++ b/autoload/clap/impl.vim
@@ -93,6 +93,11 @@ function! s:on_typed_sync_impl() abort
 
   call g:clap.display.set_lines_lazy(lines)
 
+  " vim need this to make the current selection visible.
+  if !s:is_nvim
+    call clap#sign#toggle_cursorline()
+  endif
+
   call g:clap#display_win.compact_if_undersize()
   call clap#spinner#set_idle()
```

```vim
set nocompatible
set runtimepath=~/.vim/plugged/vim-clap
set runtimepath^=~/.vim/plugged/space-vim-theme
syntax on
filetype plugin indent on

color space_vim_theme
```

It looks like this when using `:Clap files`

<img width="1138" alt="屏幕快照 2019-11-19 下午10 41 57" src="https://user-images.githubusercontent.com/8850248/69156077-d8f34480-0b1d-11ea-9d5e-96f2a78d6f06.png">

**Comment by habamax** [comment]
> But the `<C-K>` and `<C-J>` works well on my side

They works for me -- you can see list is scrolled, it is just current line is not selected.

Let me try again with minimal setup.

**Comment by habamax** [comment]
> Let me try again with minimal setup.

Tried it again (with default colorscheme) current line is not selected, list can be scrolled

**Comment by habamax** [comment]
> ```diff
> +  " vim need this to make the current selection visible.
> +  if !s:is_nvim
> +    call clap#sign#toggle_cursorline()
> +  endif
> +
> ```

Hm, I should have added those lines, shouldn't I?

Let me try again

**Comment by habamax** [comment]
> Hm, I should have added those lines, shouldn't I?
> 
> Let me try again

didn't help

**Comment by liuchengxu** [comment]
Can't reproduce this issue on macOS.

**Comment by habamax** [comment]
I have tried it in neovim-qt (windows) it works as it should:
![image](https://user-images.githubusercontent.com/234774/69158914-2a88d880-0af8-11ea-8ef7-bbcf19a0a4d7.png)

**Comment by liuchengxu** [comment]
If you would like to try fixing it on your own, see https://github.com/liuchengxu/vim-clap/blob/ccccd3a162c4cabaa3896398a50fae344b3efdc9/autoload/clap/popup.vim#L411-L414

https://github.com/liuchengxu/vim-clap/blob/ccccd3a162c4cabaa3896398a50fae344b3efdc9/autoload/clap/handler.vim#L117-L121

The navigation is done with `normal!` moves. You can try if changing the `cursorline` of the popup in `s:navigate()` helps, see `:h popup_create-arguments`.

```vim
call popup_setoptions(g:clap.display.winid, {'cursorline': 1})
```

**Comment by habamax** [comment]
spent an hour on it without success :(

Basically, all highlights exist, `sign_place` works for a regular buffer, I "can feel" that current selection is changed when I press `C-k,j` (list is scrolled when I go past last visible line).

![image](https://user-images.githubusercontent.com/234774/69216943-74b49d00-0b7e-11ea-98f3-6604080bbde0.png)

The thing is -- I don't know how to apply this `sign_place` function to a Clap generated popup window, to see if the issue is there. There is `g:clap.display.bufnr` but how to re-use it for that purpose?

I have also tried
![image](https://user-images.githubusercontent.com/234774/69217182-120fd100-0b7f-11ea-9170-01367418ec36.png)

But I doubt it should work

**Comment by habamax** [comment]
To illustrate it:

1. I have created a function and a timer triggered in 5 sec
2. function places Clap signs, for the current buffer and for the Clap buffer 
3. I can see sign in current buffer but not in Clap

![vim-clap-sign-issue](https://user-images.githubusercontent.com/234774/69217579-096bca80-0b80-11ea-9348-1d1fc3d08c91.gif)

**Comment by liuchengxu** [comment]
The point is vim's popup is not a normal window like neovim, which is unfocuable and you have to redraw manually to see the effect sometimes. See this:

https://github.com/liuchengxu/vim-clap/blob/ccccd3a162c4cabaa3896398a50fae344b3efdc9/autoload/clap/handler.vim#L119-L120

**Comment by liuchengxu** [comment]
```diff
diff --git a/autoload/clap/handler.vim b/autoload/clap/handler.vim
index d540d65..373d6db 100644
--- a/autoload/clap/handler.vim
+++ b/autoload/clap/handler.vim
@@ -118,6 +118,8 @@ else
     call s:navigate(a:direction)
     " redraw is neccessary!
     redraw
+
+    echom  "Current signs of clap buffer:".string(sign_getplaced(g:clap.display.bufnr))
     call s:on_move_safe()
   endfunction
```

See if the sign already exists when you press `<C-K>` and `<C-J>`.

**Comment by habamax** [comment]
> 
> 
> The point is vim's popup is not a normal window like neovim, which is unfocuable and you have to redraw manually to see the effect sometimes. See this:
> 
> https://github.com/liuchengxu/vim-clap/blob/ccccd3a162c4cabaa3896398a50fae344b3efdc9/autoload/clap/handler.vim#L119-L120

I have added `redraw` to the func, didn't help.

**Comment by habamax** [comment]
> See if the sign already exists when you press `<C-K>` and `<C-J>`.

Nothing in messages

**Comment by liuchengxu** [comment]
I have no idea why the messages are not recorded.

```diff
diff --git a/autoload/clap/handler.vim b/autoload/clap/handler.vim
index d540d65..421e1a4 100644
--- a/autoload/clap/handler.vim
+++ b/autoload/clap/handler.vim
@@ -118,6 +118,12 @@ else
     call s:navigate(a:direction)
     " redraw is neccessary!
     redraw
+
+    if !exists('g:sign')
+      let g:sign = []
+    endif
+    call add(g:sign, a:direction.": ".string(sign_getplaced(g:clap.display.bufnr)))
+
     call s:on_move_safe()
   endfunction

diff --git a/autoload/clap/sign.vim b/autoload/clap/sign.vim
index f165ac2..9b607ee 100644
--- a/autoload/clap/sign.vim
+++ b/autoload/clap/sign.vim
@@ -48,6 +48,11 @@ function! clap#sign#toggle_cursorline() abort
   let curlnum = g:clap.display.getcurlnum()
   call s:place_cur_sign_at(curlnum)
   let s:last_signed_id = curlnum
+
+  if !exists('g:sign')
+    let g:sign = []
+  endif
+  call add(g:sign, "toggle_cursorline: ".string(sign_getplaced(g:clap.display.bufnr)))
 endfunction
```

Then `:echo g:sign`.

The interesting thing is `g:sign` seeminly indicates no signs, but the sign does visually work as expected on my side.

**Comment by habamax** [comment]
> 
> 
> ```diff
> diff --git a/autoload/clap/handler.vim b/autoload/clap/handler.vim
> index d540d65..373d6db 100644
> --- a/autoload/clap/handler.vim
> +++ b/autoload/clap/handler.vim
> @@ -118,6 +118,8 @@ else
>      call s:navigate(a:direction)
>      " redraw is neccessary!
>      redraw
> +
> +    echom  "Current signs of clap buffer:".string(sign_getplaced(g:clap.display.bufnr))
>      call s:on_move_safe()
>    endfunction
> ```
> 
> See if the sign already exists when you press `<C-K>` and `<C-J>`.

So, it looks like `ehcom` doesn't work in winexecute context (tried different variations).

I end up using global variable `g:message` and then echoed it:

```
Current signs of clap buffer:[{'signs': [], 'bufnr': 8}]
```

**Comment by habamax** [comment]
> Then `:echo g:sign`.

```
['toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'up: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]', 'toggle_cursorline: [{''signs'': [], ''bufnr'': 5}]', 'down: [{''signs'': [], ''bufnr'': 5}]']
```

**Comment by liuchengxu** [comment]
That's the problem. The signs should not be empty.

**Comment by liuchengxu** [comment]
Before I can get access to a Windows machine, this issue will probably make no progess. If you have another time and interests, I recommend you checking around `autoload/handler.vim`, `autoload/popup.vim` and `autoload/sign.vim` to figure it out :).

**Comment by habamax** [comment]
NP, will try it

**Comment by habamax** [comment]
I have tried to reproduce sign issue with barebone popup.

For buffer of popup window `sign_place` has no effect at all:

![vim-popup-sign-issue](https://user-images.githubusercontent.com/234774/69253779-3c38b180-0bc6-11ea-9925-acce8d18c0e4.gif)

Not sure if this is a bug or expected behavior of vim's popup windows.

@Yggdroot somehow does selection on popups though. Probably he does use smth else, like vim's builtin. Cursorline?

**Comment by Yggdroot** [comment]
@habamax Perhaps it's because the default color of cursorline is too similar with the popup background so that you can't distinguish between them with eyes.

**Comment by liuchengxu** [comment]
@Yggdroot  I don't think so in that the SignColumn did not even show up.

@habamax I believe you could try reporting this minimal vimrc to https://github.com/vim/vim.

**Comment by arashm** [comment]
I have the same issue on both Linux and Mac. I downgraded the vim to version `8.1.2292` and on that version both sign and highlight works fine. The next version I have locally is `8.1.2321` which the popup doesn't show the sign or highlight.

There are 29 versions between those, but somewhere in between something has changed that caused this. I'm not sure but from 8.1.2295 looks like related from [its release notes](https://github.com/vim/vim/releases/tag/v8.1.2295). I'll try to see if I can compile vim locally and narrow it down.

**Comment by arashm** [comment]
I just compiled both `8.1.2295` and `8.1.2294`, 2294 is fine but in 2295 the sign and highlights doesn't show up. So it should be something about [this commit](https://github.com/StephenWall/vim/pull/1060/commits/4eb7dae255abc271cf313d4c75839577f1424183).

**Comment by habamax** [comment]
> I just compiled both `8.1.2295` and `8.1.2294`, 2294 is fine but in 2295 the sign and highlights doesn't show up. So it should be something about [this commit](https://github.com/StephenWall/vim/pull/1060/commits/4eb7dae255abc271cf313d4c75839577f1424183).

Thx! I have created a bug report
